qa/L0_logging/log_format_test.py

```diff
@@ -0,0 +1,470 @@
+
+
+def split_row(row):
+    return [r.strip() for r in row.group("row").strip().split("|")]
+
+
+def validate_protobuf(protobuf):
+    # Note currently we only check for model config
+    # but technically any protubuf should be valid
+
+    google.protobuf.text_format.ParseLines(
+        protobuf, grpcclient.model_config_pb2.ModelConfig()
+    )
+
+
+def validate_table(table_rows):
+    index = 0
+    top_border = table_border_regex.search(table_rows[index])
+    assert top_border
+
+    index += 1
+    header = table_row_regex.search(table_rows[index])
+    assert header
+    header = split_row(header)
+
+    index += 1
+    middle_border = table_border_regex.search(table_rows[index])
+    assert middle_border
+
+    # Process each row
+    index += 1
+    parsed_rows = []
+    row = ""
+    for index, row in enumerate(table_rows[index:]):
+        matched = table_row_regex.search(row)
+        if matched:
+            row_data = split_row(matched)
+            parsed_rows.append(row_data)
+
+    end_border = table_border_regex.search(row)
+    assert end_border
+
+    for row in parsed_rows:
+        assert len(row) == len(header)
+
+
+@validator
+def validate_message(message, escaped):
+    split_message = message.split("\n")
+    heading = split_message[0]
+    obj = split_message[1:] if len(split_message) > 1 else []
+    if heading and escaped:
+        try:
+            json.loads(heading)
+        except Exception as e:
+            raise Exception(
+                f"{e.__class__.__name__} {e}\nFirst line of message in log record is not a valid JSON string"
+            )
+    elif heading:
+        with pytest.raises(json.JSONDecodeError):
+            json.loads(heading)
+    if obj:
+        match = table_border_regex.search(obj[0])
+        if match:
+            validate_table(obj)
+        elif escaped:
+            validate_protobuf(obj)
+        else:
+            # if not escaped we can't
+            # guarantee why type of object is present
+            pass
+
+
+class TestLogFormat:
+    @pytest.fixture(autouse=True)
+    def _setup(self, request):
+        test_case_name = request.node.name
+        self._server_options = {}
+        self._server_options["log-verbose"] = INT32_MAX
+        self._server_options["log-info"] = 1
+        self._server_options["log-error"] = 1
+        self._server_options["log-warning"] = 1
+        self._server_options["log-format"] = "default"
+        self._server_options["model-repository"] = os.path.abspath(
+            os.path.join(module_directory, "log_models")
+        )
+        self._server_process = None
+        self._server_options["log-file"] = os.path.join(
+            test_logs_directory, test_case_name + ".server.log"
+        )
+
+    def _shutdown_server(self):
+        if self._server_process:
+            self._server_process.kill()
+            self._server_process.wait()
+
+    def _launch_server(self, escaped=None):
+        cmd = ["tritonserver"]
```

Comment:
follow-up, why bash process in python? Why not one of the following:
1. Start server in `test.sh` like other tests
2. Use in-process python API within python test
3. Do it this way with subprocess in python, but move it into a common util we can start to re-use in other tests if reducing bash usage.

Reply:
I'm thinking #3 - reason to do it here was to easily name the logs to match the test cases - and to also control escaping / non escaping via environment variable. Would have been difficult to reuse the start server / seperate out the cases there.

in process api is an option - but log injection was specifically a network issue - so wanted to also stay true to the original intent - I haven't tried to verify if the same holds true if accessing the in process api directly